src/Evolution/Systems/GeneralizedHarmonic/DampedHarmonicGaugeQuantities.cpp

```diff
@@ -0,0 +1,928 @@
+ *
+ * \details  Functions to compute \f$ log(g^p/N) \f$ and its spacetime
+ *  derivatives:
+ * \f[
+ * \partial_a log(g^p/N) = (p/g)\partial_a g - (1/N)\partial_a N
+ * \f]
+ */
+template <typename DataType>
+Scalar<DataType> log_factor(const Scalar<DataType>& lapse,
+    const Scalar<DataType>& sqrt_det_spatial_metric,
+    const double& exponent) noexcept {
+  auto logfac = make_with_value<Scalar<DataType>>(get(lapse), 0.);
+  if (exponent == 0.) {
+    get(logfac) = log(1. / get(lapse));
+  } else if (exponent == 0.5) {
+    get(logfac) = log(get(sqrt_det_spatial_metric) / get(lapse));
+  } else {
+    get(logfac) = log(pow(get(sqrt_det_spatial_metric),
+      2. * exponent) / get(lapse));
+  }
+  return logfac;
+}
+
+namespace {
+template <size_t SpatialDim, typename Frame, typename DataType>
+struct D4LogFacBuffer;
+
+template <size_t SpatialDim, typename Frame>
+struct D4LogFacBuffer<SpatialDim, Frame, double> {
+  // NOLINTNEXTLINE(cppcoreguidelines-pro-type-member-init)
+  explicit D4LogFacBuffer(const size_t /*size*/) noexcept {}
+
+  tnsr::a<double, SpatialDim, Frame> d_g;
+  Scalar<double> d0_N;
+  tnsr::i<double, SpatialDim, Frame> d3_N;
+  tnsr::a<double, SpatialDim, Frame> d_N;
+};
+
+template <size_t SpatialDim, typename Frame>
+struct D4LogFacBuffer<SpatialDim, Frame, DataVector> {
+  private:
+    // We make one giant allocation so that we don't thrash the heap.
+    Variables<tmpl::list<::Tags::Tempa<0, SpatialDim, Frame>,
+                        ::Tags::TempScalar<1>,
+                        ::Tags::Tempi<2, SpatialDim, Frame>,
+                        ::Tags::Tempa<3, SpatialDim, Frame>>> buffer_;
+
+  public:
+    explicit D4LogFacBuffer(const size_t size) noexcept : buffer_(size),
+      d_g(get<::Tags::Tempa<0, SpatialDim, Frame>>(buffer_)),
+      d0_N(get<::Tags::TempScalar<1>>(buffer_)),
+      d3_N(get<::Tags::Tempi<2, SpatialDim, Frame>>(buffer_)),
+      d_N(get<::Tags::Tempa<3, SpatialDim, Frame>>(buffer_)) {}
+
+    tnsr::a<DataVector, SpatialDim, Frame>& d_g;
+    Scalar<DataVector>& d0_N;
+    tnsr::i<DataVector, SpatialDim, Frame>& d3_N;
+    tnsr::a<DataVector, SpatialDim, Frame>& d_N;
+};
+}  // namespace
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> spacetime_deriv_of_log_factor(
+    const Scalar<DataType>& lapse,
+    const tnsr::I<DataType, SpatialDim, Frame>& shift,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_unit_normal,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const Scalar<DataType>& sqrt_det_spatial_metric,
+    const tnsr::ii<DataType, SpatialDim, Frame>& dt_spatial_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const double& exponent) noexcept {
+  // Use a Variables to reduce total number of allocations. This is especially
+  // important in a multithreaded environment.
```

Comment:
There seem to be only three tensor allocations in this function. Do you have an idea whether this improves the performance in any significant way?

Reply:
I haven't profiled for improvement with and without the buffer, so I don't know how much of an effect this has on performance. However, how can it hurt?